opt_einsum/contract.py

```diff
@@ -128,7 +128,7 @@ def contract_path(*operands, **kwargs):
 
     # Get length of each unique dimension and ensure all dimensions are correct
     dimension_dict = {}
-    broadcast_indices = [[] for x in range(len(input_list))]
+    bcast = set()
```

Comment:
In this approach I believe we always keep a given index as broadcast even if it is "expanded" in a given contraction. This could lead to performance regression.

Reply:
I don't think I quite understand this, do you have an example? I *think* what I did was just a refactor, apart from the case when an index is dimension 1 everywhere it appears (and is thus not broadcast), however there are many different possibilities with ``einsum`` so could have missed something!

Reply:
OK I think I see - essentially, we need to identify if an initially 'broadcasted' index becomes 'normal' at some point? I'll take a look at this and your solution in a bit!

Reply:
So nearly all the logic can be handled by just relying on ``can_blas`` already (since it was modified a few PRs ago), including the above example. In fact, if I completely remove ``bcast`` logic from ``contract_path`` then all but 1 tests still pass! The missing case (``test_singleton_dimension_broadcast``) is essentially this:
```python
np.einsum("ij,jk->ik", p, p)  # p.shape = (1, 5)
``` 
which is something I wasn't aware ``numpy.einsum`` could do. My suggestion is therefore just to give ``can_blas`` a bit of extra information to catch this case, probably the dimensions involved.

Reply:
Yes, that particular use case surprised me as well. The multitude of `np.einsum` edge cases is quite extensive.

Now that we have const expressions involved it likely does make more sense to move the logic to the `can_blas` section. Simply check that none of the shapes is 1 in the `removed_indices`.

I can add this later, but best to add the above example in tests by pulling out the einsum run data and checking if can_blas is True.

Reply:
OK I've updated it so that the logic is in``can_blas`` which also allows 'healed' broadcast indices to blas. 

Just N.B doing this made me realize that the path algorithms assume broadcast indices are their full size I think, which is not necessarily true.